client.go

```diff
@@ -0,0 +1,196 @@
+// Package iconik API for Golang
+package iconik
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+)
+
+const (
+	iconikHost     = "https://app.iconik.io/API/"
+	searchEndpoint = "search/v1/search/"
+)
+
+// Credentials are the identification required by the Iconik API
+//
+// The AppID is the application key id that you
+// get when generating an application key.
+//
+// The Token is the string generated alongside the
+// application key.
+type Credentials struct {
+	AppID string
+	Token string
+}
+
+// IClient implements a Iconik client. Do not modify state concurrently.
+type IClient struct {
+	Credentials
+
+	// If true, don't retry requests if authorization has expired
+	NoRetry bool
+
+	// If true, display debugging information about API calls
+	Debug bool
+
+	// State
+	host       string
+	httpClient http.Client
+}
+
+// NewIClient creates a new Client for accessing the Iconik API.
+func NewIClient(creds Credentials, host string) (*IClient, error) {
+	if host == "" {
+		host = iconikHost
+	}
+	c := &IClient{
+		Credentials: creds,
+		host:        iconikHost,
+	}
+
+	return c, nil
+}
+
+// Create an authorized request using the client's credentials
+func (c *IClient) newRequest(method, apiPath string, body io.Reader) (*http.Request, error) {
+	path := c.host + apiPath
+	header := make(http.Header)
+	header.Add("accept", "application/json")
+	header.Add("App-Id", c.AppID)
+	header.Add("Content-Type", "application/json")
+	header.Add("Auth-Token", c.Token)
+
+	req, err := http.NewRequest(method, path, body)
+	req.Header = header
+
+	if err != nil {
+		return nil, err
+	}
+
+	if c.Debug {
+		log.Printf("newRequest: %s %s\n", method, req.URL)
+	}
+
+	return req, nil
+}
+
+// Dispatch an authorized API GET request
+func (c *IClient) get(apiPath string, body io.Reader, header http.Header) (*http.Response, error) {
+
+	req, err := c.newRequest(http.MethodGet, apiPath, body)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := c.httpClient.Do(req)
+	return resp, err
+}
+
+func (c *IClient) post(apiPath string, body io.Reader) (*http.Response, error) {
+	req, err := c.newRequest(http.MethodPost, apiPath, body)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := c.httpClient.Do(req)
+	return resp, err
+}
+
+// Looks for an error message in the response body and parses it into a
+// IError object
+func (c *IClient) parseError(body []byte) error {
+	err := &IError{}
+	if json.Unmarshal(body, err) != nil {
+		return nil
```

Comment:
this is really weird: if the unmarshal doesn't work and there is an error, we return nil? my suggestion is to get rid of the parseError() function and just do the code in whatever was calling it.

Reply:
ah true, removed function + added the code into the parseResponse